File: client-ts/test/attest.test.ts
/**
 * These run against fixtures.json, which is produced by the Go `attest` package
 * (`go run ./client-ts/test/gen-fixtures`).
 *
 * That is the point of the file. Two implementations of a digest, each written
 * from the same description, agree with the description and not necessarily with
 * each other — and a digest that differs by one byte does not fail loudly, it
 * makes every genuine attestation look forged, so a caller that turned
 * verification on would reject every wallet the vault ever creates. Checking
 * against the Go implementation's OUTPUT is what catches that.
 */
import { strict as assert } from 'node:assert';
import { readFileSync } from 'node:fs';
import * as path from 'node:path';
import { describe, it } from 'node:test';

import {
  type Attestation,
  type NodeKey,
  type Statement,
  parseNodeKeys,
  statementDigest,
  validateNodeKeys,
  verifyAttestations,
} from '../src/attest';

interface JsonStatement {
  walletId: string;
  address: string;
  groupPubkey: string;
  threshold: number;
  allIds: number[];
}

interface Fixtures {
  statement: JsonStatement;
  digest: string;
  nodes: { partyId: number; pubKey: string }[];
  threshold: number;
  genuine: { partyId: number; signature: string }[];
  forged: { partyId: number; signature: string }[];
  replayed: { partyId: number; signature: string }[];
  mutatedStatements: { name: string; statement: JsonStatement }[];
  reorderedIds: JsonStatement;
  duplicateKeyNodes: { partyId: number; pubKey: string }[];
  selfSignedUnderTwoIds: { partyId: number; signature: string }[];
}

const fixtures: Fixtures = JSON.parse(
  readFileSync(path.join(__dirname, 'fixtures.json'), 'utf8'),
);

const stmt = (s: JsonStatement): Statement => ({
  walletId: s.walletId,
  address: s.address,
  groupPubkey: s.groupPubkey,
  threshold: s.threshold,
  allIds: s.allIds,
});
const atts = (a: { partyId: number; signature: string }[]): Attestation[] => a;
const nodes = (n: { partyId: number; pubKey: string }[]): NodeKey[] => n;

const statement = stmt(fixtures.statement);
const configured = nodes(fixtures.nodes);
const threshold = fixtures.threshold;

describe('statement digest', () => {
  it('reproduces the digest the Go implementation signs', () => {
    assert.equal(Buffer.from(statementDigest(statement)).toString('hex'), fixtures.digest);
  });

  it('does not depend on party-id order: the same group in a different order is the same group', () => {
    assert.equal(
      Buffer.from(statementDigest(stmt(fixtures.reorderedIds))).toString('hex'),
      fixtures.digest,
    );
  });

  it('sorts party ids numerically, not lexicographically', () => {
    // The default JavaScript sort would put 10 before 2 and produce a digest no
    // signer ever signed — and, being a plausible-looking 32 bytes, it would
    // fail as "unattested wallet" rather than as the bug it is.
    const ascending = { ...statement, allIds: [2, 10, 3] };
    const shuffled = { ...statement, allIds: [10, 3, 2] };
    assert.equal(
      Buffer.from(statementDigest(ascending)).toString('hex'),
      Buffer.from(statementDigest(shuffled)).toString('hex'),
    );
  });

  it('ignores address casing, so a checksummed address is not a second wallet', () => {
    const lowered = { ...statement, address: statement.address.toLowerCase() };
    assert.equal(Buffer.from(statementDigest(lowered)).toString('hex'), fixtures.digest);
  });

  it('length-prefixes fields so a boundary between two of them cannot be moved', () => {
    const a = { ...statement, walletId: 'ab', address: '0xc1' };
    const b = { ...statement, walletId: 'a', address: '0xbc1' };
    assert.notEqual(
      Buffer.from(statementDigest(a)).toString('hex'),
      Buffer.from(statementDigest(b)).toString('hex'),
    );
  });

  it('refuses a half-specified statement rather than hashing it', () => {
    const cases: [string, Statement][] = [
      ['no wallet id', { ...statement, walletId: '' }],
      ['no address', { ...statement, address: '' }],
      ['short group key', { ...statement, groupPubkey: '02'.repeat(16) }],
      ['threshold below 2', { ...statement, threshold: 1 }],
      ['fewer parties than threshold', { ...statement, threshold: 3, allIds: [1, 2] }],
    ];
    for (const [name, bad] of cases) {
      assert.throws(() => statementDigest(bad), Error, `${name} produced a digest`);
    }
  });
});

describe('verification', () => {
  it('accepts a threshold of genuine attestations', () => {
    verifyAttestations(statement, atts(fixtures.genuine), configured, threshold);
  });

  // The attack this closes: a frontend that skipped the ceremony and generated a
  // key it knows can sign for the address, but it cannot produce attestations
  // from the nodes' identity keys.
  it('rejects attestations signed by an unconfigured key', () => {
    assert.throws(
      () => verifyAttestations(statement, atts(fixtures.forged), configured, threshold),
      /below its threshold/,
    );
  });

  // One node's attestation replayed under several ids must count once, or a
  // single compromised node satisfies the threshold by itself. Here each replay
  // is caught by the recovered key not matching the id it claims.
  it('counts one node replayed under three party ids once', () => {
    assert.throws(
      () => verifyAttestations(statement, atts(fixtures.replayed), configured, threshold),
      /below its threshold/,
    );
  });

  // The same replay under the id it genuinely belongs to. Nothing about this
  // signature is invalid, so only counting DISTINCT parties stops it — a tally
  // of valid signatures would reach two and let one node clear a 2-of-3.
  it('counts one party submitted twice once', () => {
    assert.throws(
      () =>
        verifyAttestations(
          statement,
          atts([fixtures.genuine[0], fixtures.genuine[0]]),
          configured,
          threshold,
        ),
      /below its threshold/,
    );
  });

  it('binds the attestation to every field of the statement', () => {
    for (const m of fixtures.mutatedStatements) {
      assert.throws(
        () => verifyAttestations(stmt(m.statement), atts(fixtures.genuine), configured, threshold),
        Error,
        `genuine attestations verified against a statement with a mutated ${m.name}`,
      );
    }
  });

  it('rejects an attestation from a party the caller does not know', () => {
    const unknown = configured.filter((n) => n.partyId !== 2);
    assert.throws(
      () => verifyAttestations(statement, atts(fixtures.genuine), unknown, threshold),
      /below its threshold/,
    );
  });

  it('fails closed with no configured node keys', () => {
    assert.throws(
      () => verifyAttestations(statement, atts(fixtures.genuine), [], threshold),
      /no node identity keys configured/,
    );
  });

  it('refuses a threshold below 2', () => {
    assert.throws(
      () => verifyAttestations(statement, atts(fixtures.genuine), configured, 1),
      /below 2/,
    );
  });

  it('refuses one node configured under two party ids, before counting with it', () => {
    const dup = nodes(fixtures.duplicateKeyNodes);
    assert.throws(
      () =>
        verifyAttestations(statement, atts(fixtures.selfSignedUnderTwoIds), dup, threshold),
      /same public key/,
    );
    assert.throws(() => validateNodeKeys(dup), /same public key/);
  });
});

describe('node key configuration', () => {
  it('accepts a well-formed set', () => {
    validateNodeKeys(configured);
  });

  it('rejects an empty set, a repeated party id, and a truncated key', () => {
    assert.throws(() => validateNodeKeys([]), /no node identity keys configured/);
    assert.throws(
      () => validateNodeKeys([configured[0], { partyId: 1, pubKey: configured[1].pubKey }]),
      /configured more than once/,
    );
    assert.throws(
      () =>
        validateNodeKeys([
          configured[0],
          { partyId: 2, pubKey: (configured[1].pubKey as string).slice(0, 20) },
        ]),
      /33 compressed bytes/,
    );
  });

  it('parses a configuration string and verifies with the result', () => {
    const spec = fixtures.nodes.map((n) => `${n.partyId}:${n.pubKey}`).join(',');
    verifyAttestations(statement, atts(fixtures.genuine), parseNodeKeys(spec), threshold);
  });

  it('refuses an empty configuration string rather than silently disabling verification', () => {
    assert.throws(() => parseNodeKeys('  '), /would disable attestation verification/);
    assert.throws(() => parseNodeKeys('1'), /partyId:publicKeyHex/);
  });
});

File: attest/attest_test.go
package attest

import (
	"encoding/hex"
	"strings"
	"testing"

	"github.com/ethereum/go-ethereum/crypto"
)

func newNode(t *testing.T, partyID int) (string, NodeKey) {
	t.Helper()
	key, err := crypto.GenerateKey()
	if err != nil {
		t.Fatal(err)
	}
	return hex.EncodeToString(crypto.FromECDSA(key)), NodeKey{
		PartyID: partyID,
		PubKey:  crypto.CompressPubkey(&key.PublicKey),
	}
}

func testStatement() Statement {
	return Statement{
		WalletID:  "wallet-1",
		Address:   "0x1111111111111111111111111111111111111111",
		GroupPub:  make([]byte, 33),
		Threshold: 2,
		AllIDs:    []int{1, 2, 3},
	}
}

// The honest path: t of the configured nodes attest, and the client accepts.
func TestThresholdOfConfiguredNodesVerifies(t *testing.T) {
	k1, n1 := newNode(t, 1)
	k2, n2 := newNode(t, 2)
	_, n3 := newNode(t, 3)
	s := testStatement()

	a1, err := Sign(k1, 1, s)
	if err != nil {
		t.Fatal(err)
	}
	a2, err := Sign(k2, 2, s)
	if err != nil {
		t.Fatal(err)
	}
	if err := Verify(s, []*Attestation{a1, a2}, []NodeKey{n1, n2, n3}, 2); err != nil {
		t.Fatalf("two genuine attestations were rejected: %v", err)
	}
}

// The attack this closes: a frontend that skipped the ceremony and generated a
// key it knows can sign for the address, but it cannot produce attestations from
// the nodes' identity keys.
func TestFrontendCannotForgeAttestations(t *testing.T) {
	_, n1 := newNode(t, 1)
	_, n2 := newNode(t, 2)
	_, n3 := newNode(t, 3)
	s := testStatement()

	// The frontend holds some key of its own and signs with it, claiming to be
	// each party in turn.
	rogueHex, _ := newNode(t, 0)
	var forged []*Attestation
	for _, id := range []int{1, 2, 3} {
		a, err := Sign(rogueHex, id, s)
		if err != nil {
			t.Fatal(err)
		}
		forged = append(forged, a)
	}
	if err := Verify(s, forged, []NodeKey{n1, n2, n3}, 2); err == nil {
		t.Fatal("attestations signed by an unconfigured key were accepted")
	}
}

// One node's attestation replayed under several party ids must count once. Any
// other behaviour would let a single compromised node satisfy the threshold.
func TestReplayedAttestationCountsOnce(t *testing.T) {
	k1, n1 := newNode(t, 1)
	_, n2 := newNode(t, 2)
	_, n3 := newNode(t, 3)
	s := testStatement()

	a1, err := Sign(k1, 1, s)
	if err != nil {
		t.Fatal(err)
	}
	// Same signature, presented three times, twice under other parties' ids.
	dup := []*Attestation{
		a1,
		{PartyID: 2, Signature: a1.Signature},
		{PartyID: 3, Signature: a1.Signature},
	}
	if err := Verify(s, dup, []NodeKey{n1, n2, n3}, 2); err == nil {
		t.Fatal("one node's attestation replayed under other ids satisfied the threshold")
	}
}

// An attestation is bound to its exact statement. Changing any field must
// invalidate it, or a frontend could attest to one wallet and hand over another.
func TestAttestationIsBoundToEveryField(t *testing.T) {
	k1, n1 := newNode(t, 1)
	k2, n2 := newNode(t, 2)
	s := testStatement()
	a1, _ := Sign(k1, 1, s)
	a2, _ := Sign(k2, 2, s)
	nodes := []NodeKey{n1, n2}

	mutations := map[string]func(Statement) Statement{
		"wallet id": func(x Statement) Statement { x.WalletID = "other"; return x },
		"address": func(x Statement) Statement {
			x.Address = "0x2222222222222222222222222222222222222222"
			return x
		},
		"group key": func(x Statement) Statement {
			pub := make([]byte, 33)
			pub[0] = 0x02
			x.GroupPub = pub
			return x
		},
		"threshold":   func(x Statement) Statement { x.Threshold = 3; return x },
		"party set":   func(x Statement) Statement { x.AllIDs = []int{1, 2, 4}; return x },
		"party count": func(x Statement) Statement { x.AllIDs = []int{1, 2}; return x },
	}
	for name, mutate := range mutations {
		t.Run(name, func(t *testing.T) {
			if err := Verify(mutate(s), []*Attestation{a1, a2}, nodes, 2); err == nil {
				t.Fatal("attestations verified against a different statement")
			}
		})
	}
}

// Party-id order must not change the digest: the same group described in a
// different order is the same group.
func TestPartyOrderDoesNotChangeTheDigest(t *testing.T) {
	a := testStatement()
	b := testStatement()
	b.AllIDs = []int{3, 1, 2}
	da, err := a.Digest()
	if err != nil {
		t.Fatal(err)
	}
	db, err := b.Digest()
	if err != nil {
		t.Fatal(err)
	}
	if da != db {
		t.Fatal("reordering party ids changed the attestation digest")
	}
}

// Fields are length-prefixed so a boundary between two of them cannot be moved.
func TestFieldBoundariesCannotBeShifted(t *testing.T) {
	a := testStatement()
	a.WalletID = "ab"
	a.Address = "0xc111111111111111111111111111111111111111"
	b := testStatement()
	b.WalletID = "a"
	b.Address = "0xbc11111111111111111111111111111111111111"
	da, _ := a.Digest()
	db, _ := b.Digest()
	if da == db {
		t.Fatal("moving the boundary between two fields produced the same digest")
	}
}

// A client with no configured node keys must fail closed. Accepting attestations
// it cannot check would mean trusting the frontend's word — the exact thing this
// package exists to stop.
func TestNoConfiguredNodesFailsClosed(t *testing.T) {
	k1, _ := newNode(t, 1)
	s := testStatement()
	a1, _ := Sign(k1, 1, s)
	if err := Verify(s, []*Attestation{a1}, nil, 2); err == nil {
		t.Fatal("verification passed with no configured node keys")
	}
	if err := Verify(s, []*Attestation{a1}, nil, 2); !strings.Contains(err.Error(), "out of band") &&
		!strings.Contains(err.Error(), "no node identity keys") {
		t.Fatalf("error does not explain the missing configuration: %v", err)
	}
}

// Malformed statements must not produce a digest at all — signing over a
// half-specified assertion would attest to something nobody can check.
func TestMalformedStatementsAreRefused(t *testing.T) {
	cases := map[string]Statement{
		"no wallet id":                 {Address: "0x1", GroupPub: make([]byte, 33), Threshold: 2, AllIDs: []int{1, 2}},
		"no address":                   {WalletID: "w", GroupPub: make([]byte, 33), Threshold: 2, AllIDs: []int{1, 2}},
		"short group key":              {WalletID: "w", Address: "0x1", GroupPub: make([]byte, 32), Threshold: 2, AllIDs: []int{1, 2}},
		"threshold below 2":            {WalletID: "w", Address: "0x1", GroupPub: make([]byte, 33), Threshold: 1, AllIDs: []int{1, 2}},
		"fewer parties than threshold": {WalletID: "w", Address: "0x1", GroupPub: make([]byte, 33), Threshold: 3, AllIDs: []int{1, 2}},
	}
	for name, s := range cases {
		t.Run(name, func(t *testing.T) {
			if _, err := s.Digest(); err == nil {
				t.Fatal("a malformed statement produced a digest")
			}
		})
	}
}

// The address comparison is case-insensitive, so a checksummed address and its
// lowercase form attest to the same wallet rather than silently to two.
func TestAddressCaseDoesNotMatter(t *testing.T) {
	a := testStatement()
	a.Address = "0xAbCdEf1111111111111111111111111111111111"
	b := testStatement()
	b.Address = strings.ToLower(a.Address)
	da, _ := a.Digest()
	db, _ := b.Digest()
	if da != db {
		t.Fatal("address casing changed the attestation digest")
	}
}

// A single node configured under two party ids must not satisfy a 2-of-n
// threshold by itself. Verify counts distinct party ids, so if a config file
// repeats one public key under two entries — a copy-paste, which nothing else
// would flag — that one node signs twice, is counted twice, and the threshold
// that was supposed to require two independent hosts requires one.
func TestDuplicatePublicKeyUnderTwoPartyIDsIsRejected(t *testing.T) {
	k1, n1 := newNode(t, 1)
	_, n3 := newNode(t, 3)
	// Party 2 is misconfigured with party 1's public key.
	n2 := NodeKey{PartyID: 2, PubKey: n1.PubKey}
	s := testStatement()

	// The single node holding that key attests under both ids.
	a1, err := Sign(k1, 1, s)
	if err != nil {
		t.Fatal(err)
	}
	a2, err := Sign(k1, 2, s)
	if err != nil {
		t.Fatal(err)
	}
	if err := Verify(s, []*Attestation{a1, a2}, []NodeKey{n1, n2, n3}, 2); err == nil {
		t.Fatal("one node satisfied a 2-of-3 threshold because its key was configured under two party ids")
	}

	// And the same set must be rejected at configuration time, before any
	// attestation is presented against it.
	if err := ValidateNodeKeys([]NodeKey{n1, n2, n3}); err == nil {
		t.Fatal("ValidateNodeKeys accepted one public key configured under two party ids")
	}
}

// The rest of the configuration contract: a party id may name exactly one node,
// keys must be well formed, and an empty set is not a valid configuration.
func TestValidateNodeKeys(t *testing.T) {
	_, n1 := newNode(t, 1)
	_, n2 := newNode(t, 2)

	if err := ValidateNodeKeys([]NodeKey{n1, n2}); err != nil {
		t.Fatalf("a well-formed node set was rejected: %v", err)
	}
	if err := ValidateNodeKeys(nil); err == nil {
		t.Fatal("an empty node set was accepted")
	}
	if err := ValidateNodeKeys([]NodeKey{n1, {PartyID: 1, PubKey: n2.PubKey}}); err == nil {
		t.Fatal("party id 1 was accepted twice")
	}
	if err := ValidateNodeKeys([]NodeKey{n1, {PartyID: 2, PubKey: n2.PubKey[:32]}}); err == nil {
		t.Fatal("a truncated public key was accepted")
	}
}
